### PR TITLE
Add 3D secure functionality

### PR DESCRIPTION
**Pull request by steve-gr4vy**
# Description

Adding support for 3D Secure flows

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own changes
- [ ] I have run `yarn test` to make sure my changes pass all linters
- [ ] I have pulled the latest changes from the upstream `main` branch

## Contribution guidelines

For contribution guidelines, styleguide, and other helpful information please
see the `CONTRIBUTING.md` file in the root of this project.

### CODE REVIEW

File: reference/openapi.yaml [hunk 1]
```diff
@@ -295,12 +295,6 @@ paths:
     post:
       $ref: "./paths/transactions__capture-transaction.yaml"
 
-  "/transactions/{transaction_id}/authorize":
-    parameters:
-      - $ref: "./path-parameters/transaction_id.yaml"
-    post:
-      $ref: "./paths/transactions__authorize-transaction.yaml"
-
```
Comment (cbetta) [inline reference/openapi.yaml:303]: Tnx for removing this.

File: reference/paths/transactions__post-approve-transaction.yaml [hunk 2]
```diff
@@ -1,10 +1,22 @@
...
+    application/json:
+      schema:
+        $ref: "../openapi.yaml#/components/schemas/TransactionApproveRequest"
+      examples:
+        Fingerprint step complete:
+          value:
+            
+        Challenge step complete:
+          value:
+            threeds_transaction_status: "N"
```
Comment (cbetta) [inline reference/paths/transactions__post-approve-transaction.yaml:16]: Is this `N` just an example value?

File: reference/request-bodies/threeds_data_request.yaml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  - eci
+  - cavv
+  - directory_response
+  - authentication_response
+
+properties:
+  version:
+    type: string
+    description: |-
+      The version of 3D Secure used.
```
Comment (cbetta) [inline reference/request-bodies/threeds_data_request.yaml]:
```suggestion
      The version of 3-D Secure used. 
```

File: reference/openapi.yaml [hunk 4]
```diff
@@ -441,6 +439,8 @@ components:
     #   $ref: "./resources/transaction_event.yaml"
     # TransactionEvents:
     #   $ref: "./resources/transaction_events.yaml"
+    ThreeDSDataRequest:
```
Comment (cbetta) [inline reference/openapi.yaml:442]: Let's stick to 3-D Secure and `3d_secure`

Comment (cbetta) [reply reference/openapi.yaml:442]:
```suggestion
    3DSecureData:
```

File: reference/paths/transactions__post-approve-transaction.yaml (hunk 2 above)
Comment (steve-gr4vy) [reply reference/paths/transactions__post-approve-transaction.yaml:16]: Yes, whatever is passed back from the issuer in the cRes should be passed in this field, however cardinal's examples show an N or Y in this field, hence the example value of 'N'

File: reference/openapi.yaml (hunk 4 above)
Comment (cbetta) [reply reference/openapi.yaml:442]: Actually, this does not need to be its own schema. We should just define this straight in the card request schema.

File: reference/request-bodies/transaction_approve_request.yaml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+---
+title: Approve a transaction
+type: object
+description: A request to capture a transaction.
+
+x-tags:
+  - Request Bodies
+
+properties:
+  threeds_transaction_status:
```
Comment (cbetta) [inline reference/request-bodies/transaction_approve_request.yaml:10]:
```suggestion
  3d_secure_transaction_status:
```

File: reference/request-bodies/transaction_approve_request.yaml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+---
+title: Approve a transaction
+type: object
+description: A request to capture a transaction.
```
Comment (cbetta) [inline reference/request-bodies/transaction_approve_request.yaml:4]:
```suggestion
description: A request to approve the 3D secure step of a card transaction.
```

File: reference/request-bodies/transaction_approve_request.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+---
+title: Approve a transaction
```
Comment (cbetta) [inline reference/request-bodies/transaction_approve_request.yaml:2]:
```suggestion
title: Complete a 3D Secure transaction
```

File: reference/request-bodies/transaction_create_request.yaml [hunk 8]
```diff
@@ -39,6 +39,20 @@ properties:
       The optional payment method details to create an authorization for. This
       field is required for processing a card.
 
+  threeds_data:
```
Comment (cbetta) [inline reference/request-bodies/transaction_create_request.yaml:42]:
```suggestion
  3d_secure_data:
```

File: reference/request-bodies/transaction_create_request.yaml [hunk 9]
```diff
@@ -39,6 +39,20 @@ properties:
...
+    x-model-name: TransactionThreeDSecureDataRequest
```
Comment (cbetta) [inline reference/request-bodies/transaction_create_request.yaml:43]: Not needed

File: reference/request-bodies/transaction_create_request.yaml [hunk 10]
```diff
@@ -39,6 +39,20 @@ properties:
...
+    $ref: "../openapi.yaml#/components/schemas/ThreeDSDataRequest"
```
Comment (cbetta) [inline reference/request-bodies/transaction_create_request.yaml:44]: I'd just embed this data in this schema.

File: reference/resources/transaction.yaml [hunk 11]
```diff
@@ -131,3 +132,10 @@ properties:
       - production
     description: |-
       The environment this transaction has been created in.
+
+  threeds_included:
+    type: boolean
+    description: |-
+      Whether the 3DS flow was included in the transaction.
+    example: false
+    nullable: true
```
Comment (cbetta) [inline reference/resources/transaction.yaml:141]: Hmm. Not sure what we want to capture here. That a transaction had 3D Secure performed? What if it failed? What if it hasn't completed yet?

**Review by cbetta — CHANGES_REQUESTED** [review]

File: reference/paths/transactions__post-approve-transaction.yaml (hunk 2 above)
Comment (cbetta) [reply reference/paths/transactions__post-approve-transaction.yaml:16]: Cool. My only change would be to update this to:

```suggestion
            3d_secure_transaction_status: "N"
```

File: reference/request-bodies/threeds_data_request.yaml [hunk 12]
```diff
@@ -1,8 +1,8 @@
 ---
-title: External 3-D Secure data
+title: External `3-D` Secure data
 type: object
 description: |-
-  Data from an external 3-D Secure MPI that can be used to successfully create the transaction.
+  Data from an external `3-D` Secure MPI that can be used to successfully create the transaction.
```
Comment (cbetta) [inline reference/request-bodies/threeds_data_request.yaml:5]: Why `3-D`? Spell checker getting in the way?

Comment (steve-gr4vy) [reply reference/request-bodies/threeds_data_request.yaml:5]: Yes, spell checker didn't like 3-D, hence the backticks

Comment (cbetta) [reply reference/request-bodies/threeds_data_request.yaml:5]: Ok. A nicer way to resolve this is to add the term to the `.spelling` file.

**Comment by douglaseggleton** [comment]
Closing this PR in favour of https://github.com/gr4vy/gr4vy-openapi/pull/65
(A lot has changed so thought it was better to just create a new PR and copy the stuff thats needed across)

**Pull request closed without merging**